File: packages/pushduck-py/pushduck/multipart.py
"""Multipart uploads: provider operations and session tokens.

The four S3 calls are made with ``urllib.request`` and parsed with
``xml.etree``, both stdlib, so this package still installs nothing. Pulling in
boto3 for four requests would be larger than everything else here combined.

Three provider behaviours below are not obvious from the API documentation. They
were learnt expensively in the TypeScript implementation and reproduced here
rather than rediscovered, which is the whole point of having a shared protocol:

* S3 can return an error document with HTTP 200. ``CompleteMultipartUpload``
  streams its response, so it commits to a status before it knows whether the
  operation succeeded. Checking only the status treats a failure as success.
* ``ListParts`` is paginated, and a truncated listing looks exactly like a
  complete one — which makes a resuming client re-upload parts the provider
  already holds.
* Aborting an upload that is already gone returns 404, which is the desired end
  state rather than an error.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import urllib.error
import urllib.request
import xml.etree.ElementTree as ElementTree
from typing import Any, Dict, List, Mapping, Optional, Tuple

from .config import UploadConfig
from .errors import UploadError
from .sign import presign

#: S3 namespaces every element in these responses.
_NS = {"s3": "http://s3.amazonaws.com/doc/2006-03-01/"}


def _text(element: Optional[ElementTree.Element], path: str) -> str:
    if element is None:
        return ""
    # Try namespaced first, then bare: MinIO and AWS agree, but some
    # S3-compatible providers omit the namespace entirely.
    found = element.find(f"s3:{path}", _NS)
    if found is None:
        found = element.find(path)
    return (found.text or "").strip() if found is not None else ""


def _provider_error(operation: str, code: str, message: str) -> UploadError:
    """Map a provider code onto a protocol code.

    The provider's own code stays in the message because that is what a
    developer searches for; the protocol code is what a client branches on.
    """
    if code in ("AccessDenied", "InvalidAccessKeyId", "SignatureDoesNotMatch"):
        return UploadError(
            "STORAGE_ACCESS_DENIED", f"{operation} denied by storage ({code}): {message}"
        )
    if code == "NoSuchUpload":
        return UploadError(
            "NOT_FOUND", f"{operation} failed: the multipart session no longer exists"
        )
    if code in ("EntityTooSmall", "InvalidPart", "InvalidPartOrder"):
        return UploadError("BAD_REQUEST", f"{operation} rejected ({code}): {message}")
    return UploadError("STORAGE_UNAVAILABLE", f"{operation} failed ({code}): {message}")


def _signed_request(
    config: UploadConfig,
    method: str,
    key: str,
    *,
    query: Optional[Mapping[str, str]] = None,
    body: bytes = b"",
    headers: Optional[Mapping[str, str]] = None,
    operation: str = "storage request",
) -> bytes:
    """Sign, send, and return the body — converting any failure into an UploadError."""
    scheme, host, path = config.object_address(key)

    url = presign(
        access_key_id=config.access_key_id,
        secret_access_key=config.secret_access_key,
        session_token=config.session_token,
        method=method,
        scheme=scheme,
        host=host,
        path=path,
        region=config.region,
        query=query,
        headers=headers,
        expires_in=900,
        now=config.now(),
    )

    request = urllib.request.Request(url, data=body or None, method=method)
    for name, value in (headers or {}).items():
        request.add_header(name, value)

    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            payload = response.read()
            status = response.status
    except urllib.error.HTTPError as error:
        payload = error.read()
        status = error.code
    except OSError as error:  # DNS, refused connection, TLS
        raise UploadError("STORAGE_UNAVAILABLE", "Could not reach storage") from error

    # An error document can arrive with a 200, so the body is inspected
    # regardless of status.
    if b"<Error" in payload:
        try:
            root = ElementTree.fromstring(payload)
            code = _text(root, "Code")
            if code:
                raise _provider_error(operation, code, _text(root, "Message"))
        except ElementTree.ParseError:
            pass

    if status >= 400:
        raise _provider_error(operation, f"HTTP{status}", payload.decode("utf-8", "replace")[:200])

    return payload


def create_multipart_upload(config: UploadConfig, key: str, content_type: str) -> str:
    """Start a session and return the provider's upload id."""
    headers = {"x-amz-acl": "private"}
    if content_type:
        headers["content-type"] = content_type

    payload = _signed_request(
        config,
        "POST",
        key,
        query={"uploads": ""},
        headers=headers,
        operation="CreateMultipartUpload",
    )

    upload_id = _text(ElementTree.fromstring(payload), "UploadId")
    if not upload_id:
        raise UploadError("STORAGE_UNAVAILABLE", "CreateMultipartUpload returned no upload id")

    return upload_id


def presign_upload_part(
    config: UploadConfig, key: str, upload_id: str, part_number: int
) -> str:
    """Return a URL the browser can PUT one part to.

    ``partNumber`` and ``uploadId`` go into the canonical request *before*
    signing. Appending them afterwards signs a different URL than the one used.
    """
    scheme, host, path = config.object_address(key)

    return presign(
        access_key_id=config.access_key_id,
        secret_access_key=config.secret_access_key,
        session_token=config.session_token,
        method="PUT",
        scheme=scheme,
        host=host,
        path=path,
        region=config.region,
        query={"partNumber": str(part_number), "uploadId": upload_id},
        expires_in=config.upload_expiry,
        now=config.now(),
    )


def complete_multipart_upload(
    config: UploadConfig, key: str, upload_id: str, parts: List[Dict[str, Any]]
) -> None:
    """Stitch the parts into an object."""
    # Providers reject an unordered list, and the client is not required to send
    # them in order.
    ordered = sorted(parts, key=lambda part: int(part.get("partNumber", 0)))

    document = ["<CompleteMultipartUpload>"]
    for part in ordered:
        # The ETag is echoed exactly as the provider gave it, quotes included.
        # Stripping them produces `InvalidPart` after every byte has already
        # been transferred.
        etag = str(part.get("etag", "")).replace("&", "&amp;").replace("<", "&lt;")
        document.append(
            f"<Part><PartNumber>{int(part.get('partNumber', 0))}</PartNumber>"
            f"<ETag>{etag}</ETag></Part>"
        )
    document.append("</CompleteMultipartUpload>")

    _signed_request(
        config,
        "POST",
        key,
        query={"uploadId": upload_id},
        body="".join(document).encode("utf-8"),
        headers={"content-type": "application/xml"},
        operation="CompleteMultipartUpload",
    )


def abort_multipart_upload(config: UploadConfig, key: str, upload_id: str) -> None:
    """Discard a session and its parts.

    Abandoned parts are billed until removed and do not appear in a normal
    object listing, so this runs on any permanent failure.
    """
    try:
        _signed_request(
            config,
            "DELETE",
            key,
            query={"uploadId": upload_id},
            operation="AbortMultipartUpload",
        )
    except UploadError as error:
        # Already gone is the desired end state, not a failure.
        if error.code == "NOT_FOUND":
            return
        raise


def list_uploaded_parts(
    config: UploadConfig, key: str, upload_id: str
) -> List[Dict[str, Any]]:
    """Report which parts the provider actually holds.

    This is what makes resume possible: the client's own record is a hint, and
    the provider is the authority.
    """
    parts: List[Dict[str, Any]] = []
    marker = ""

    while True:
        query = {"uploadId": upload_id}
        if marker:
            query["part-number-marker"] = marker

        payload = _signed_request(
            config, "GET", key, query=query, operation="ListParts"
        )

        try:
            root = ElementTree.fromstring(payload)
        except ElementTree.ParseError as error:
            raise UploadError("STORAGE_UNAVAILABLE", "ListParts returned malformed XML") from error

        found = root.findall("s3:Part", _NS) or root.findall("Part")
        for element in found:
            etag = _text(element, "ETag")
            # A part with no ETag cannot be completed with, so including it
            # would produce `InvalidPart` much later.
            if not etag:
                continue
            parts.append({"partNumber": int(_text(element, "PartNumber") or 0), "etag": etag})

        # A truncated listing is indistinguishable from a complete one unless
        # this flag is honoured, and stopping early makes a resuming client
        # re-upload parts the provider already has.
        truncated = _text(root, "IsTruncated").lower() == "true"
        next_marker = _text(root, "NextPartNumberMarker")
        if not truncated or not next_marker:
            return parts
        marker = next_marker


# ─── session tokens ──────────────────────────────────────────────────────────


def _session_signature(secret: str, payload: str) -> str:
    signature = hmac.new(
        f"pushduck-multipart:{secret}".encode("utf-8"),
        payload.encode("utf-8"),
        hashlib.sha256,
    ).digest()
    return base64.urlsafe_b64encode(signature).rstrip(b"=").decode("ascii")


def sign_session(secret: str, session: Mapping[str, Any]) -> str:
    """Issue a token binding a key, upload id and route.

    Not a convenience. If a client sent ``{key, uploadId}`` directly, anyone who
    guessed or observed that pair could sign parts for — or abort — another
    user's upload. Middleware authenticates the caller; nothing would tie the
    caller to the object.
    """
    claim = json.dumps(dict(session), separators=(",", ":"), sort_keys=True)
    payload = base64.urlsafe_b64encode(claim.encode("utf-8")).rstrip(b"=").decode("ascii")
    return f"{payload}.{_session_signature(secret, payload)}"


def verify_session(secret: str, token: Any) -> Dict[str, Any]:
    """Verify a token and return the session it asserts.

    One message for every failure mode, so a probe learns nothing about which
    part was wrong.
    """
    rejection = UploadError("FORBIDDEN", "Invalid or expired multipart session")

    if not isinstance(token, str):
        raise rejection

    parts = token.split(".")
    if len(parts) != 2 or not parts[0] or not parts[1]:
        raise rejection

    if not hmac.compare_digest(parts[1], _session_signature(secret, parts[0])):
        raise rejection

    try:
        padded = parts[0] + "=" * (-len(parts[0]) % 4)
        session = json.loads(base64.urlsafe_b64decode(padded))
    except (ValueError, UnicodeDecodeError) as error:
        raise rejection from error

    # A valid signature over a malformed payload must still not proceed.
    if (
        not isinstance(session, dict)
        or not session.get("key")
        or not session.get("uploadId")
        or int(session.get("partSize") or 0) <= 0
    ):
        raise rejection

    return session

File: packages/pushduck-go/pushduck/multipart.go
package pushduck

// Multipart session tokens and the four multipart actions.
//
// A multipart upload spans four requests, and the last three have to name the
// object they act on. If the client sent `{key, uploadId}` directly, anyone
// could sign parts for — or abort — someone else's upload by guessing or
// observing that pair. Route metadata hooks authenticate the *caller*; nothing
// would tie the caller to the object.
//
// So `init` returns an opaque token binding the key, upload id and route under
// an HMAC, and later calls present the token instead of raw identifiers. The
// server re-derives the values rather than trusting anything the client sent.
// Byte-compatible with the TypeScript implementation, including the secret
// prefix, so the two remain interchangeable.

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"strings"
)

// multipartSession is what a token asserts once verified.
type multipartSession struct {
	Key       string `json:"key"`
	UploadID  string `json:"uploadId"`
	Route     string `json:"route"`
	PartSize  int64  `json:"partSize"`
	TotalSize int64  `json:"totalSize"`
}

func (router *Router) sessionSignature(payload string) string {
	mac := hmac.New(sha256.New, []byte("pushduck-multipart:"+router.config.SecretAccessKey))
	mac.Write([]byte(payload))
	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

func (router *Router) signSession(session multipartSession) (string, error) {
	claim, err := json.Marshal(session)
	if err != nil {
		return "", NewError("INTERNAL_ERROR", "Could not encode multipart session")
	}
	payload := base64.RawURLEncoding.EncodeToString(claim)
	return payload + "." + router.sessionSignature(payload), nil
}

func (router *Router) verifySession(token string) (multipartSession, error) {
	// One message for every failure mode, so a probe learns nothing about
	// which part was wrong.
	reject := func() (multipartSession, error) {
		return multipartSession{}, NewError("FORBIDDEN", "Invalid or expired multipart session")
	}

	parts := strings.Split(token, ".")
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return reject()
	}
	if !hmac.Equal([]byte(parts[1]), []byte(router.sessionSignature(parts[0]))) {
		return reject()
	}

	raw, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return reject()
	}

	var session multipartSession
	if err := json.Unmarshal(raw, &session); err != nil {
		return reject()
	}
	// A valid signature over a malformed payload must still not proceed.
	if session.Key == "" || session.UploadID == "" || session.PartSize <= 0 {
		return reject()
	}

	return session, nil
}

// authorizeSession verifies the token *and* re-runs the route's metadata hooks.
//
// Both are required. The token proves which object is being acted on; the
// hooks prove the caller is still allowed to act. Checking only the token
// would let a revoked user finish an upload they started.
func (router *Router) authorizeSession(
	r *http.Request, routeName string, route Route, token string,
) (multipartSession, error) {
	session, err := router.verifySession(token)
	if err != nil {
		return multipartSession{}, err
	}

	if session.Route != routeName {
		return multipartSession{}, NewError("FORBIDDEN", "Invalid or expired multipart session")
	}

	for _, hook := range route.Metadata {
		if _, err := hook(r, FileMeta{Name: session.Key}); err != nil {
			return multipartSession{}, err
		}
	}

	return session, nil
}

// ─── limits ─────────────────────────────────────────────────────────────────

const (
	mib = 1 << 20
	// The intersection of every supported provider's limits.
	minPartSize     = 5 * mib
	maxPartSize     = 5 * 1024 * mib
	maxParts        = 10_000
	defaultPartSize = 5 * mib
)

// choosePartSize clamps a requested size into the provider-safe range and
// raises it when the 10,000-part cap would otherwise be exceeded.
//
// Parts are uniform except the last. That is stricter than S3 requires and
// deliberate: Cloudflare R2 rejects an upload whose non-final parts differ in
// size, so uniform parts are the only sizing that works everywhere.
func choosePartSize(totalSize, requested int64) int64 {
	size := requested
	if size <= 0 {
		size = defaultPartSize
	}
	if size < minPartSize {
		size = minPartSize
	}
	if size > maxPartSize {
		size = maxPartSize
	}

	for totalSize > size*maxParts {
		size *= 2
		if size > maxPartSize {
			return maxPartSize
		}
	}

	return size
}

// ─── actions ────────────────────────────────────────────────────────────────

type multipartInitRequest struct {
	File     FileMeta       `json:"file"`
	Metadata map[string]any `json:"metadata"`
	PartSize int64          `json:"partSize"`
}

type multipartSignRequest struct {
	Session     string `json:"session"`
	PartNumbers []int  `json:"partNumbers"`
}

type multipartCompleteRequest struct {
	Session  string          `json:"session"`
	Parts    []CompletedPart `json:"parts"`
	File     FileMeta        `json:"file"`
	Metadata map[string]any  `json:"metadata"`
}

type multipartSessionRequest struct {
	Session string `json:"session"`
}

func (router *Router) multipartInit(
	w http.ResponseWriter, r *http.Request, routeName string, route Route,
	telemetry map[string]string,
) error {
	var request multipartInitRequest
	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
		return NewError("BAD_REQUEST", "Request body must be JSON")
	}
	if request.File.Size <= 0 {
		return NewError("BAD_REQUEST",
			"`file` with name, size and type is required to start a multipart upload")
	}

	// Metadata hooks and validation run here exactly as they do for a single PUT,
	// so a multipart upload cannot bypass a route's constraints.
	metadata := map[string]any{}
	for key, value := range request.Metadata {
		metadata[key] = value
	}
	for _, hook := range route.Metadata {
		produced, err := hook(r, request.File)
		if err != nil {
			return err
		}
		if produced != nil {
			metadata = produced
		}
	}

	if message := route.validate(request.File); message != "" {
		return NewError("VALIDATION_FAILED", message)
	}

	key := generateKey(request.File.Name)
	uploadID, err := router.config.createMultipartUpload(key, request.File.Type)
	if err != nil {
		return err
	}

	partSize := choosePartSize(request.File.Size, request.PartSize)

	token, err := router.signSession(multipartSession{
		Key: key, UploadID: uploadID, Route: routeName,
		PartSize: partSize, TotalSize: request.File.Size,
	})
	if err != nil {
		// The session exists at the provider but cannot be handed out, so it
		// would be billed forever with nobody able to finish or abort it.
		_ = router.config.abortMultipartUpload(key, uploadID)
		return err
	}

	writeJSON(w, http.StatusOK, telemetry, map[string]any{
		"success": true, "session": token, "key": key,
		"partSize": partSize, "metadata": metadata,
	})
	return nil
}

func (router *Router) multipartSign(
	w http.ResponseWriter, r *http.Request, routeName string, route Route,
	telemetry map[string]string,
) error {
	var request multipartSignRequest
	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
		return NewError("BAD_REQUEST", "Request body must be JSON")
	}
	if request.PartNumbers == nil {
		return NewError("BAD_REQUEST", "`partNumbers` must be an array")
	}

	session, err := router.authorizeSession(r, routeName, route, request.Session)
	if err != nil {
		return err
	}

	maxPart := int((session.TotalSize + session.PartSize - 1) / session.PartSize)
	if maxPart < 1 {
		maxPart = 1
	}

	signed := make([]map[string]any, 0, len(request.PartNumbers))
	for _, partNumber := range request.PartNumbers {
		// A part number outside the plan would sign a write past the end of
		// the object the session was created for.
		if partNumber < 1 || partNumber > maxPart {
			return NewError("BAD_REQUEST",
				"Part number is outside this upload")
		}

		start := int64(partNumber-1) * session.PartSize
		size := session.PartSize
		if remaining := session.TotalSize - start; remaining < size {
			size = remaining
		}

		signed = append(signed, map[string]any{
			"partNumber": partNumber,
			"url":        router.config.presignUploadPart(session.Key, session.UploadID, partNumber),
			"size":       size,
		})
	}

	writeJSON(w, http.StatusOK, telemetry, signed)
	return nil
}

func (router *Router) multipartComplete(
	w http.ResponseWriter, r *http.Request, routeName string, route Route,
	telemetry map[string]string,
) error {
	var request multipartCompleteRequest
	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
		return NewError("BAD_REQUEST", "Request body must be JSON")
	}
	if request.Parts == nil {
		return NewError("BAD_REQUEST", "`parts` must be an array")
	}

	session, err := router.authorizeSession(r, routeName, route, request.Session)
	if err != nil {
		return err
	}

	if err := router.config.completeMultipartUpload(session.Key, session.UploadID, request.Parts); err != nil {
		return err
	}

	if route.OnComplete != nil {
		metadata := map[string]any{}
		for key, value := range request.Metadata {
			metadata[key] = value
		}
		if err := route.OnComplete(r, session.Key, request.File, metadata); err != nil {
			return err
		}
	}

	host, path := router.config.objectAddress(session.Key)
	writeJSON(w, http.StatusOK, telemetry, map[string]any{
		"success": true, "key": session.Key, "url": "https://" + host + path,
	})
	return nil
}

func (router *Router) multipartAbort(
	w http.ResponseWriter, r *http.Request, routeName string, route Route,
	telemetry map[string]string,
) error {
	var request multipartSessionRequest
	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
		return NewError("BAD_REQUEST", "Request body must be JSON")
	}

	session, err := router.authorizeSession(r, routeName, route, request.Session)
	if err != nil {
		return err
	}

	if err := router.config.abortMultipartUpload(session.Key, session.UploadID); err != nil {
		return err
	}

	writeJSON(w, http.StatusOK, telemetry, map[string]any{"success": true})
	return nil
}

func (router *Router) multipartParts(
	w http.ResponseWriter, r *http.Request, routeName string, route Route,
	telemetry map[string]string,
) error {
	var request multipartSessionRequest
	if err := json.NewDecoder(r.Body).Decode(&request); err != nil {
		return NewError("BAD_REQUEST", "Request body must be JSON")
	}

	session, err := router.authorizeSession(r, routeName, route, request.Session)
	if err != nil {
		return err
	}

	parts, err := router.config.listUploadedParts(session.Key, session.UploadID)
	if err != nil {
		return err
	}

	writeJSON(w, http.StatusOK, telemetry, map[string]any{
		"success": true, "parts": parts,
	})
	return nil
}
